pact/marmalade.repl

```diff
@@ -1,2 +1,348 @@
+
+  (expect-failure "create a token with token-id that doesn't satisfy t: protocol"
+    "Token protocol violation"
+    (create-token
+      (create-token-id { 'uri: "test-uri-wrong", 'precision: 0, 'policies: [] } )
+      0 "test-uri" [] ))
+
+  (expect-failure "create a token with uri starting with marmalade fails"
+    "Reserved protocol: marmalade:"
+    (create-token
+      (create-token-id { 'uri: "marmalade:uri", 'precision: 0, 'policies: [] } )
+       0 "marmalade:uri" [] ))
+
+  (expect "create a token with token-id that satisfies t: protocol"
+    true
+    (create-token
+      (create-token-id { 'uri: "test-uri", 'precision: 0, 'policies: [] } )
+       0 "test-uri" []))
+
+  (expect "Token info is inserted into table"
+    { "id": "t:YxyMWZItsZ_bxQgq7PzzR8RaD_xDxi1zlIOqGRoler4"
+     ,"supply": 0.0
+     ,"precision": 0
+     ,"uri": "test-uri"
+     ,"policies": []
+    }
+    (get-token-info "t:YxyMWZItsZ_bxQgq7PzzR8RaD_xDxi1zlIOqGRoler4")
+  )
+
+  (expect "create-token events"
+     (format "{}" [[
+     {"name": "marmalade-v2.ledger.TOKEN","params": ["t:YxyMWZItsZ_bxQgq7PzzR8RaD_xDxi1zlIOqGRoler4" 0 0.0 [] "test-uri"]}]])
+     (format "{}" [(map (remove "module-hash")  (env-events true))])
+  )
+
+(commit-tx)
+
+(begin-tx "mint")
+  (use marmalade-v2.ledger)
+
+  (env-data {
+     "token-id": "t:YxyMWZItsZ_bxQgq7PzzR8RaD_xDxi1zlIOqGRoler4"
+    ,"account": "k:account"
+    ,"account-guard": {"keys": ["account"], "pred": "keys-all"}
+    })
+
+  (expect-failure "mint fails without MINT capability in scope"
+    "Managed capability not installed: (marmalade-v2.ledger.MINT"
+    (mint (read-msg 'token-id) (read-msg 'account) (read-keyset 'account-guard) 1.0))
+
+  (env-sigs [
+    { 'key: 'any
+     ,'caps: [(MINT (read-msg "token-id") "k:account" 1.0)]}
+   ])
+
+  (expect "mint succeeds"
+    true
+    (mint (read-msg 'token-id) (read-msg 'account) (read-keyset 'account-guard) 1.0))
+
+  (expect "Token supply is updated"
+    { "id": (read-string "token-id")
+     ,"supply": 1.0
+     ,"precision": 0
+     ,"uri": "test-uri"
+     ,"policies": []
+    }
+    (get-token-info (read-string "token-id"))
+  )
+
+  (expect "Account is credited"
+    1.0
+    (get-balance (read-string "token-id") (read-string "account"))
+  )
+
+  (expect "mint events"
+    [
+      {"name": "marmalade-v2.ledger.MINT","params": [(read-string "token-id") (read-string "account") 1.0]}
+      {"name": "marmalade-v2.ledger.ACCOUNT_GUARD","params": [(read-string "token-id") (read-string "account") (read-keyset "account-guard")]}
+      {"name": "marmalade-v2.ledger.RECONCILE","params": [(read-string "token-id") 1.0 {"account": "","current": 0.0,"previous": 0.0} {"account": "k:account","current": 1.0,"previous": 0.0}]}
+      {"name": "marmalade-v2.ledger.SUPPLY","params": [(read-string "token-id") 1.0]}
+    ]
+    (map (remove "module-hash")  (env-events true))
+  )
+
+(commit-tx)
+
+(begin-tx "offer, withdraw with timeout")
```

Comment:
Might be good to add a case trying to withdraw before the timeout period has passed?

Reply:
Nice one that was missing, added.